pkg/counter/cache.go

```diff
@@ -175,24 +203,82 @@ func indexOfMember(members []toolchainv1alpha1.Member, name string) int {
+	if toolchainStatus.Status.HostOperator == nil ||
+		toolchainStatus.Status.HostOperator.MasterUserRecordCount == 0 {
+		return initializeFromResources(cl, toolchainStatus.Namespace)
+	}
+	// otherwise, initialize the cached counters from the ToolchainStatud resource.
+	return initializeFromToolchainStatus(cl, toolchainStatus)
 }
 
-func loadCurrentResources(cl client.Client, namespace string) error {
+// initialize the cached counters from the UserSignup and MasterUserRecord resources.
+// this func lists all UserSignup and MasterUserRecord resources
+func initializeFromResources(cl client.Client, namespace string) error {
+	log.Info("initializing counters from resources")
+	usersignups := &toolchainv1alpha1.UserSignupList{}
+	if err := cl.List(context.TODO(), usersignups, client.InNamespace(namespace)); err != nil {
+		return err
+	}
 	murs := &toolchainv1alpha1.MasterUserRecordList{}
 	if err := cl.List(context.TODO(), murs, client.InNamespace(namespace)); err != nil {
 		return err
 	}
 	reset()
+	for _, usersignup := range usersignups.Items {
+		if value, exists := usersignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey]; exists {
+			_, err := strconv.Atoi(value) // let's make sure the value is actually an integer
+			if err != nil {
+				log.Error(err, "invalid number of activations", "name", usersignup.Name, "value", value)
+				continue
+			}
+			cachedCounts.UsersPerActivationCounts[value]++
+		}
+	}
 	for _, mur := range murs.Items {
 		cachedCounts.MasterUserRecordCount++
 		for _, ua := range mur.Spec.UserAccounts {
 			cachedCounts.UserAccountsPerClusterCounts[ua.TargetCluster]++
 		}
 	}
 	cachedCounts.initialized = true
+	log.Info("cached counts initialized from UserSignups and MasterUserRecords",
+		"MasterUserRecordCount", cachedCounts.MasterUserRecordCount,
+		"UserAccountsPerClusterCounts", cachedCounts.UserAccountsPerClusterCounts,
+		"UsersPerActivationCounts", cachedCounts.UsersPerActivationCounts)
+	return nil
+}
+
+// initialize the cached counters from the ToolchainStatud resource.
+func initializeFromToolchainStatus(cl client.Client, toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
+	// initialize the cached counters from the ToolchainStatus resource
+	if toolchainStatus.Status.HostOperator == nil {
+		toolchainStatus.Status.HostOperator = &toolchainv1alpha1.HostOperatorStatus{}
+	}
+	// MasterUserRecordCount
+	cachedCounts.MasterUserRecordCount += toolchainStatus.Status.HostOperator.MasterUserRecordCount
+	// UserAccountsPerClusterCounts
+	for _, memberStatus := range toolchainStatus.Status.Members {
+		cachedCounts.UserAccountsPerClusterCounts[memberStatus.ClusterName] += memberStatus.UserAccountCount
+	}
+	// UsersPerActivationCounts
+	if toolchainStatus.Status.Metrics != nil {
+		if metric, exists := toolchainStatus.Status.Metrics[toolchainv1alpha1.UsersPerActivationMetricKey]; exists {
+			cachedCounts.UsersPerActivationCounts = metric
```

Comment:
I guess that we should go over all items in the map and tot up the numbers from both of the maps, otherwise, we could be losing some provisioned users every time when we restart the deployment.

Reply:
done in 86cc036